source/rpt/rpt_cell_reconstruction.cc

```diff
@@ -0,0 +1,626 @@
+{
+  // Seed the random number generator
+  srand(parameters.rpt_param.seed);
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::execute_cell_reconstruction()
+{
+  assign_detector_positions(); // Assign detector position & create detectors
+  read_counts(); // Read reconstruction counts of unknown particle positions
+
+  // Read known positions if analyse positions if enable
+  if (parameters.reconstruction_param.analyse_positions)
+    {
+      read_known_positions();
+      AssertThrow(
+        known_positions.size() == reconstruction_counts.size(),
+        ExcMessage(
+          "Reconstruction counts and known positions files do not have data for the same number of positions."))
+    }
+
+  create_grid();            // Read a grid for the reactor vessel
+  set_coarse_mesh_counts(); // Calculate counts at vertices of the coarse mesh
+  find_all_positions();
+
+  // Export positions in file
+  export_positions();
+
+  // Show information about cells and vertices
+  if (parameters.rpt_param.verbose)
+    {
+      std::cout << std::endl;
+      std::cout << "Number of active cells : " << triangulation.n_active_cells()
+                << std::endl;
+      std::cout << "Number of vertices in the mesh : "
+                << triangulation.n_vertices() << std::endl;
+      std::cout << "Number of vertices with calculated counts : "
+                << map_vertices_index.size() << std::endl;
+    }
+
+  // Generate visualisation files
+  visualize_positions();
+
+  // Disable the output of time clock
+  if (!parameters.reconstruction_param.verbose_clock)
+    computing_timer.disable_output();
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::set_coarse_mesh_counts()
+{
+  find_vertices_positions(parameters.reconstruction_param.coarse_mesh_level);
+  calculate_counts();
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::find_all_positions()
+{
+  // Show positions in the terminal if verbose is enable
+  for (unsigned int i = 0; i < reconstruction_counts.size(); i++)
+    {
+      find_unknown_position(reconstruction_counts[i], i);
+      if (parameters.rpt_param.verbose)
+        std::cout << "Unknown particle : " << i
+                  << " Cell volume : " << cells_volumes[i]
+                  << " Position : " << reconstruction_positions[i] << std::endl;
+    }
+}
+
+template <int dim>
+void
+RPTCellReconstruction<dim>::find_unknown_position(
+  std::vector<double> &particle_reconstruction_counts,
+  unsigned int         id)
+{
+  int              level = 0;
+  std::vector<int> cells_indexes, new_cells_indexes;
+  Point<dim>       reconstruction_position;
+  double           volume;
+  bool             still_new_candidates = true;
+  std::string      status;
+
+  // Find the first cell candidates from coarse mesh
+  cells_indexes = find_cells(level, particle_reconstruction_counts);
+
+  // Find the positions with the more refine candidate cell
+  while (level < parameters.reconstruction_param.reactor_refinement &&
```

Comment:
You can query the number of levels from the triangulation itself. It will be safer.
https://www.dealii.org/current/doxygen/deal.II/classTriangulation.html
 use the member method n_levels () const